ir/pattern/constraint.rs

```diff
@@ -2696,3 +2710,62 @@ impl<ID: IrID> fmt::Display for LinksDeduplication<ID> {
         write!(f, "LinksDeduplication({}, {}))", self.links1, self.links2)
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct OptimisedToUnsatisfiable {
+    conjunction: Conjunction,
+}
+
+impl OptimisedToUnsatisfiable {
+    pub(crate) fn new(conjunction: Conjunction) -> OptimisedToUnsatisfiable {
+        Self { conjunction }
+    }
+
+    pub fn ids<ID: IrID>(&self) -> impl Iterator<Item = ID> {
+        [].into_iter()
+    }
+
+    pub fn vertices<ID: IrID>(&self) -> impl Iterator<Item = &Vertex<ID>> {
+        [].into_iter()
+    }
+
+    pub fn ids_foreach<F, ID: IrID>(&self, mut function: F)
+    where
+        F: FnMut(ID),
+    {
+    }
+
+    pub fn map<ID: IrID, T: Clone>(self, mapping: &HashMap<ID, T>) -> OptimisedToUnsatisfiable {
+        self
+    }
+}
+
+impl PartialEq<Self> for OptimisedToUnsatisfiable {
+    fn eq(&self, other: &Self) -> bool {
+        self.conjunction.scope_id() == other.conjunction.scope_id()
```

Comment:
I dont think the scope ID should be in here at all - any Unsatisfiable constraint is equal to any other.  Notice how none of the other constraints care about their context/scope when doing equivalence checks.